### PR TITLE
Css carousel

### PR DESCRIPTION
**Pull request by MikeKingdom**
**This replaces PR #5741**

#### What does this PR do?
This changes the `Carousel` component fairly significantly, solving some issues that exist in the current component.

This changes:
- Transitions between slides.
- Adding dimensions to the carousel container (fixed height and width).
- New ways to navigate between slides.

#### Where should the reviewer start?
See the Carousel directory under components. See the JS files there.
#### What testing has been done on this PR?
General code cleanup and organization. Tested in SB with numerous use cases.
Interesting use cases:
- Carousel with only one slide.
- Carousel with too too many slides.
- Spamming the controls (e.g. clicking the next arrow very fast).
- Skipping around slides. Going from the last slide to the first slide with the "Next" button and vice versa.
- General accessibility. Navigating through controls and any clickable child elements.
#### How should this be manually tested?
Carousel stories in Storybook.
#### What are the relevant issues?
- #3536 
- #5675 
- Possibly others?

### CODE REVIEW

File: src/js/components/Carousel/StyledCarousel.js [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+  display: ${(props) => props.displayProp};
+  animation: ${(props) =>
+    props.animation
+      ? css`
+          ${props.animation}
+          ${props.animationDuration / 1000}s
+        ease-in-out
+        `
+      : `none`};
+  animation-fill-mode: both;
```
Comment (MikeKingdom) [inline src/js/components/Carousel/StyledCarousel.js:36]: This was added so the translate won't pop back to where it was once the animation completes

**Review by jcfilben — COMMENTED** [review]
When the `full` prop is added the container is stretching on transitions. (Testing with the Carousel/Simple story)

https://user-images.githubusercontent.com/54560994/145889988-68cb31d7-9b37-43dd-9679-066d77fb9854.mp4

**Comment by MikeKingdom** [comment]
> When the `fill` prop is added the container is stretching on transitions. (Testing with the Carousel/Simple story)
Thanks Jessica. This might be a regression. Although just adding fill to that Carousel may not make sense. It's not really inside anything that would bound it's size. A typical example like from grommet site has it's parent with some bounds. For example
```jsx
<Box height="small" width="medium" overflow="hidden">
```
That example from grommet site still works as expected.

**Review by ericsoderberghp — COMMENTED** [review]
One small comment about a comment.

What happened to the "Multi" story though? It doesn't look like it's working, compared to master.

**Comment by MikeKingdom** [comment]
> One small comment about a comment.
> 
> What happened to the "Multi" story though? It doesn't look like it's working, compared to master.

Looks like the way it hides children until needed for animation/display makes it so the auto height changes if some children are taller or shorter. I'm not sure how to solve that. Thoughts?

**Comment by ericsoderberghp** [comment]
> > One small comment about a comment.
> > What happened to the "Multi" story though? It doesn't look like it's working, compared to master.
> 
> Looks like the way it hides children until needed for animation/display makes it so the auto height changes if some children are taller or shorter. I'm not sure how to solve that. Thoughts?

I don't think we need to handle differing heights. So, I'm OK with changing the images to all be the same size.

**Comment by ericsoderberghp** [comment]
I'm not seeing the controls in the Multi story. Also, the Carousel isn't taking the full width of my screen for this story, like it does on the master one.

**Comment by MikeKingdom** [comment]
> I'm not seeing the controls in the Multi story. Also, the Carousel isn't taking the full width of my screen for this story, like it does on the master one.

The bottom controls aren't showing in Multi since the example was changed to not show them on purpose:
```jsx
<Carousel controls="arrows">
```

But the right side arrow control is sometimes way to the right, disconnected from the images because the screen is bigger than the natural size of the images and they aren't getting scaled to fill the available space beyond their natural size. This is happening because of the `float: left;` on the `div` surrounding each child. This `float` seems to be there to make animation work on the child on the right (ie. keep it next to the child on the left as it moves.) I'm not quite sure why this is required beyond just using the transformX() in the animation.

**Review by jcfilben — COMMENTED** [review]
I'm getting the following warning when trying to use the `onChild` prop
Screen Shot 2021-12-21 at 12 28 54 PM
Tested with the following code in the Carousel/Simple story `onChild={(index) =>
        console.log('Index of the element currently shown.', index)
      }`

**Review by jcfilben — COMMENTED** [review]
Just a minor comment, everything else looks good to me

File: src/js/components/Carousel/Carousel.js [hunk 2]
```diff
@@ -50,7 +67,7 @@ const Carousel = ({
       });
     }
   }
-
+*/
```
[the remark this answers is not in the document]
Comment (ericsoderberghp) [reply src/js/components/Carousel/Carousel.js]: good catch, removed

**Review by jcfilben — APPROVED** [review]

**Review by halocline — APPROVED** [review]
Looks good. No concerns, no issues.

**Pull request merged**